djstripe/models/billing.py

```diff
@@ -216,6 +216,38 @@ class PromotionCode(StripeModel):
 
     stripe_class = stripe.PromotionCode
 
+    @property
+    def promotion_code(self):
+        return self.stripe_data.get("code")
+
+    @property
+    def active(self):
+        return self.stripe_data.get("active")
+
+    @property
+    def created(self):
+        return self.stripe_data.get("created")
+
+    @property
+    def times_redeemed(self):
+        return self.stripe_data.get("times_redeemed")
+
+    @property
+    def max_redemptions(self):
+        return self.stripe_data.get("max_redemptions")
+
+    @property
+    def coupon_percent_off(self):
+        return self.stripe_data.get("coupon", {}).get("percent_off")
+
+    @property
+    def coupon_duration(self):
+        return self.stripe_data.get("coupon", {}).get("duration")
+
+    @property
+    def coupon_duration_in_months(self):
+        return self.stripe_data.get("coupon", {}).get("duration_in_months")
```

Comment:
Please remove the `coupon_` properties; what you can do is a `coupon` property which would return either a `Coupon` object or None.

This brings up the need for a helper function which does something like:

```py
def get_by_id_or_object(cls, id: str | dict):
  if isinstance(id, dict):
    if "id" not in id: raise ValueError
    id = id["id"]
    return cls.objects.filter(id=id).first()
    # Also need to handle the case where it's not in the database yet before 3.0
```

So you'd do something like:

```py
def coupon(self) -> Coupon:
    return Coupon.get_by_id_or_object(self.stripe_data["coupon"])
```

You will note it's safe to use a direct lookup for this instead of `get` because `coupon` is always an attribute of the stripe data.

Similarly, you can remove the `get` from the other attributes that are not nullable. I do think we will need to handle the possibility of a missing value there at some point, but I don't think we should be doing this invisibly at this level; instead probably trigger a resync of the object, but this is something to handle separately IMO and not in the scope of this PR.

Reply:
> Please remove the `coupon_` properties; what you can do is a `coupon` property which would return either a `Coupon` object or None.
> 
> This brings up the need for a helper function which does something like:
> 
> ```python
> def get_by_id_or_object(cls, id: str | dict):
>   if isinstance(id, dict):
>     if "id" not in id: raise ValueError
>     id = id["id"]
>     return cls.objects.filter(id=id).first()
>     # Also need to handle the case where it's not in the database yet before 3.0
> ```
> 
> So you'd do something like:
> 
> ```python
> def coupon(self) -> Coupon:
>     return Coupon.get_by_id_or_object(self.stripe_data["coupon"])
> ```
> 
> You will note it's safe to use a direct lookup for this instead of `get` because `coupon` is always an attribute of the stripe data.
> 
> Similarly, you can remove the `get` from the other attributes that are not nullable. I do think we will need to handle the possibility of a missing value there at some point, but I don't think we should be doing this invisibly at this level; instead probably trigger a resync of the object, but this is something to handle separately IMO and not in the scope of this PR.

I will leave this for another PR